lib/internal/modules/esm/loader.js

```diff
@@ -168,100 +194,246 @@ class Loader {
     const { module } = await job.run();
+
     return {
       namespace: module.getNamespace(),
     };
   }
 
-  async import(specifier, parent) {
-    const job = await this.getModuleJob(specifier, parent);
-    const { module } = await job.run();
-    return module.getNamespace();
+  async getModuleJob(specifier, parentURL) {
+    const resolution = await this.resolve(specifier, parentURL);
+    const { url } = resolution;
+    let { format } = resolution;
+    let job = this.moduleMap.get(url);
+    // CommonJS will set functions for lazy job evaluation.
+    if (typeof job === 'function') this.moduleMap.set(url, job = job());
+
+    if (job !== undefined) return job;
+
+    const moduleProvider = async (url, isMain) => {
+      const loaded = await this.load(url, { format });
+      const { source } = loaded;
+      ({ format } = loaded);
+
+      const translator = translators.get(format);
+
+      if (!translator) throw new ERR_UNKNOWN_MODULE_FORMAT(format);
+
+      return translator.call(this, url, source, isMain);
+    };
+
+    const inspectBrk = (
+      parentURL === undefined &&
+      getOptionValue('--inspect-brk')
+    );
+
+    job = new ModuleJob(
+      this,
+      url,
+      moduleProvider,
+      parentURL === undefined,
+      inspectBrk
+    );
+
+    this.moduleMap.set(url, job);
+
+    return job;
   }
 
-  hook(hooks) {
-    const {
-      resolve,
-      dynamicInstantiate,
-      getFormat,
-      getSource,
-      transformSource,
-      getGlobalPreloadCode,
-    } = hooks;
+  /**
+   *
+   * @param {string | string[]} specifiers Path(s) to the module
+   * @param {string} [parentURL] Path of the parent importing the module
+   * @returns {object | object[]} A list of module export(s)
+   */
+  async import(specifiers, parentURL) {
+    const wasArr = ArrayIsArray(specifiers);
+    if (!wasArr) specifiers = [specifiers];
 
-    // Use .bind() to avoid giving access to the Loader instance when called.
-    if (resolve !== undefined)
-      this._resolve = FunctionPrototypeBind(resolve, null);
-    if (dynamicInstantiate !== undefined) {
-      process.emitWarning(
-        'The dynamicInstantiate loader hook has been removed.');
-    }
-    if (getFormat !== undefined) {
-      this._getFormat = FunctionPrototypeBind(getFormat, null);
-    }
-    if (getSource !== undefined) {
-      this._getSource = FunctionPrototypeBind(getSource, null);
-    }
-    if (transformSource !== undefined) {
-      this._transformSource = FunctionPrototypeBind(transformSource, null);
-    }
-    if (getGlobalPreloadCode !== undefined) {
-      this._getGlobalPreloadCode =
-        FunctionPrototypeBind(getGlobalPreloadCode, null);
+    const namespaces = new Array(specifiers.length);
+
+    for (const { 0: i, 1: specifier } of specifiers.entries()) {
+      const job = await this.getModuleJob(specifier, parentURL);
+      const { module } = await job.run();
+
+      namespaces[i] = module.getNamespace();
     }
```

Comment:
This already existed; I just tweaked it with new names.

It uses WeakMap, which I assume is very much intentional. Handcuffing ourselves because users might be determined to screw themselves seems very unnecessary IMO.

Reply:
Sorry my comment wasn't clear, I was refering to the fact we should avoid the lookup of `specifiers.entries` which is user-mutable.
I wasn't sure if `specifiers` was an iterator (`Array`, `Set`, generator, etc.) or an array-like object (object that has a `length` property and integer properties). If it's an iterator, we should use `%Array.from%` instead of the `for of` loop; if it's an array-like, we should use a classic `for` loop instead of the `for of` one. Since `addCustomLoaders` expects an array-like already, we should align with that (and that way we are not relying on user-mutable hooray):
```suggestion
    const namespaces = new Array(specifiers.length);

    for (let i = 0; i <  specifiers.length; i++) {
      const specifier = specifiers[i];
      const job = await this.getModuleJob(specifier, parentURL);
      const { module } = await job.run();

      namespaces[i] = module.getNamespace();
    }
```

Reply:
@aduh95 `specifiers` is whatever we want it to be: `import` is basically an internal method: it's used directly for loading custom loaders, but I think in all other situations, it's used internally within ESMLoader. An array is probably easiest.